include/sycl/algorithm/sort.hpp

```diff
@@ -173,6 +173,787 @@ void sequential_sort(cl::sycl::queue q, cl::sycl::buffer<T, 1, Alloc> buf,
+    global_start =
+        (item.get_group(0) + (item.get_group(0) / m_stage) * m_stage) *
+            item.get_local_range(0) +
+        item.get_local_id(0);
+    global_offset = m_stage * item.get_local_range(0);
+
+    /* Perform swap */
+    input1 = m_globalBuf[global_start];
+    input2 = m_globalBuf[global_start + global_offset];
+    comp = (((input1 < input2) ^ dir) * 4 + add);
+    m_globalBuf[global_start] =
+        shuffle2(input1, input2, (comp).template as<mask_op_vec_type_>());
+    m_globalBuf[global_start + global_offset] =
+        shuffle2(input2, input1, (comp).template as<mask_op_vec_type_>());
+  }
+
+private:
+  typename bitonic_sort_base<T, U>::global_buffer_accessor_t m_globalBuf;
+  typename bitonic_sort_base<T, U>::local_buffer_accessor_t m_localBuf;
+  const unsigned int m_stage;
+  const unsigned int m_highStage;
+};
+
+/**
+ * @brief
+ *
+ */
+template <typename T, unsigned U>
+class bitonic_sort_merge : public bitonic_sort_base<T, U> {
+public:
+  bitonic_sort_merge(
+      const typename bitonic_sort_base<T, U>::global_buffer_accessor_t
+          &globalBuf,
+      const typename bitonic_sort_base<T, U>::local_buffer_accessor_t &localBuf,
+      const unsigned int stage, const int dir)
+      : m_globalBuf(globalBuf), m_localBuf(localBuf), m_stage(stage),
+        mDir(dir) {}
+
+  /**
+   * @brief
+   *
+   */
+  void operator()(cl::sycl::nd_item<1> item) {
+    using namespace cl::sycl;
+#ifdef EMULATE_SHUFFLE_BUILTINS
+    using namespace emulated_shuffle_builtins;
+#endif
+
+    typedef typename bitonic_sort_base<T, U>::data_vec_type data_vec_type_;
+    typedef
+        typename bitonic_sort_base<T, U>::mask_op_vec_type mask_op_vec_type_;
+    typedef typename bitonic_sort_base<T, U>::relational_op_vec_type
+        relational_op_vec_type_;
+    const int vec_elems_per_thread_ =
+        bitonic_sort_base<T, U>::vec_elems_per_thread_::value;
+
+    data_vec_type_ input1, input2;
+    relational_op_vec_type_ comp, add;
+    unsigned int global_start, global_offset;
+
+    add = relational_op_vec_type_(4, 5, 6, 7);
+
+    /* Determine location of data in global memory */
+    global_start =
+        (item.get_group(0) + (item.get_group(0) / m_stage) * m_stage) *
+            item.get_local_range(0) +
+        item.get_local_id(0);
+    global_offset = m_stage * item.get_local_range(0);
+
+    /* Perform swap */
+    input1 = m_globalBuf[global_start];
+    input2 = m_globalBuf[global_start + global_offset];
+    comp = ((input1 < input2 ^ mDir) * 4 + add);
+    m_globalBuf[global_start] =
+        shuffle2(input1, input2, (comp).template as<mask_op_vec_type_>());
+    m_globalBuf[global_start + global_offset] =
+        shuffle2(input2, input1, (comp).template as<mask_op_vec_type_>());
+  }
+
+private:
+  typename bitonic_sort_base<T, U>::global_buffer_accessor_t m_globalBuf;
+  typename bitonic_sort_base<T, U>::local_buffer_accessor_t m_localBuf;
+  const unsigned int m_stage;
+  const int mDir;
```

Comment:
Please choose a name that's more meaningful than "dir" (I don't know what that is).

Reply:
> Sorting direction. Zero means 'ascending', one means 'descending'.